sale_timesheet_line_exclude/models/account_analytic_line.py

```diff
@@ -0,0 +1,107 @@
+    def _onchange_task_id_employee_id(self):
+        if self.project_id:  # timesheet only
+            self.so_line = self._timesheet_get_sale_line()
+
+    @api.onchange('exclude_from_sale_order')
+    def _onchange_exclude_from_sale_order(self):
+        if self.project_id:  # timesheet only
+            self.so_line = self._timesheet_get_sale_line()
+
+    @api.multi
+    def _timesheet_get_sale_line(self):
+        self.ensure_one()
+        return self._timesheet_determine_sale_line(
+            **self._timesheet_determine_sale_line_arguments()
+        ) if not self.exclude_from_sale_order else False
+
+    @api.model
+    def _timesheet_should_evaluate_so_line(self, values, check):
+        return check([field_name in values for field_name in [
+            'task_id',
+            'employee_id',
+            'exclude_from_sale_order',
+        ]])
+
+    @api.multi
+    def _timesheet_determine_sale_line_arguments(self, values=None):
+        return {
+            'task': (
+                self.env['project.task'].sudo().browse(values['task_id'])
+            ) if values and 'task_id' in values else self.task_id,
+            'employee': (
+                self.env['hr.employee'].sudo().browse(values['employee_id'])
+            ) if values and 'employee_id' in values else self.employee_id,
+        }
+
+    @api.multi
+    @api.depends('exclude_from_sale_order')
+    def _compute_timesheet_invoice_type(self):
+        result = super()._compute_timesheet_invoice_type()
+
+        for timesheet in self:
+            # Skip non-timesheet entries
+            if not timesheet.project_id:
+                continue
+
+            # Since no-task entries are 'non_billable_project', check only
+            # entries with tasks
+            if timesheet.task_id and timesheet.exclude_from_sale_order:
+                timesheet.timesheet_invoice_type = 'non_billable'
+
+        return result
+
+    @api.model
+    def _timesheet_preprocess(self, values):
+        values = super()._timesheet_preprocess(values)
+
+        if self._timesheet_should_evaluate_so_line(values, all):
+            so_line = self._timesheet_determine_sale_line(
+                **self._timesheet_determine_sale_line_arguments(values)
+            ) if not values.get('exclude_from_sale_order') else False
+            values['so_line'] = so_line.id if so_line else False
+
+        return values
+
+    @api.multi
+    def _timesheet_postprocess_values(self, values):
+        result = super()._timesheet_postprocess_values(values)
+
+        if self._timesheet_should_evaluate_so_line(values, any):
+            for timesheet in self:
+                so_line = timesheet._timesheet_get_sale_line()
+                result[timesheet.id].update({
+                    'so_line': so_line.id if so_line else False,
+                })
+
+        return result
+
+    @api.multi
+    def write(self, values):
+        if 'exclude_from_sale_order' in values and self.filtered(
```

Comment:
Why not doing this in an `@api.constrains('exclude_from_sale_order')`?

Reply:
To be consistent with how it's done in base module, but can redo that to constraint if you think it's worth

Reply:
Yes, it's clearer